File: verba-data/src/main/java/ru/nsu/ccfit/verba/verbadata/controllers/ExceptionHandlingController.java
package ru.nsu.ccfit.verba.verbadata.controllers;

import org.springframework.context.support.DefaultMessageSourceResolvable;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseBody;
import ru.nsu.ccfit.verba.verbadata.platform.ResponseDto;

import ru.nsu.ccfit.verba.verbadata.platform.enumeration.Code;
import ru.nsu.ccfit.verba.verbadata.platform.exeption.BasicVerbaException;

import java.util.List;

@ResponseBody
@ControllerAdvice
public class ExceptionHandlingController {

    @ExceptionHandler(BasicVerbaException.class)
    public ResponseEntity<ResponseDto<Void>> handleLicenceException(BasicVerbaException exception) {
        return getResponseWithError(exception.getCode(), exception);
    }

    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<ResponseDto<Void>> handleMethodArgumentNotValidException(MethodArgumentNotValidException exception) {
        List<String> messageException =  exception.getAllErrors().stream()
                .map(DefaultMessageSourceResolvable::getDefaultMessage)
                .toList();
        return getResponseWithError(Code.BAD_REQUEST, new Exception(messageException.toString()));
    }

    @ExceptionHandler(Exception.class)
    public ResponseEntity<ResponseDto<Void>> handleError(Exception exception) {
        return getResponseWithError(Code.INTERNAL_SERVER_ERROR, exception);
    }

    private ResponseEntity<ResponseDto<Void>> getResponseWithError(Code code, Exception exception) {

        return ResponseEntity.status(code.getHttpStatus())
                .body(ResponseDto.withError(code, exception.getMessage()));
    }
}

File: verba-api/src/main/kotlin/ru/nsu/ccfit/verba/verbaapi/app/ExceptionHandlingController.kt
package ru.nsu.ccfit.verba.verbaapi.app

import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseBody
import ru.nsu.ccfit.verba.verbaapi.platform.Code
import ru.nsu.ccfit.verba.verbaapi.platform.Response
import ru.nsu.ccfit.verba.verbaapi.platform.exception.BasicVerbaException

@ResponseBody
@ControllerAdvice
class ExceptionHandlingController {

    @ExceptionHandler(BasicVerbaException::class)
    fun handleVerbaException(exception: BasicVerbaException): ResponseEntity<Response<Void>> {
        return getResponseWithError(exception.errorCode, exception)
    }

    @ExceptionHandler(Exception::class)
    fun handleError(exception: Exception): ResponseEntity<Response<Void>> {
        return getResponseWithError(Code.INTERNAL_SERVER_ERROR, exception)
    }

    private fun getResponseWithError(errorCode: Code, exception: Exception): ResponseEntity<Response<Void>> {
        return ResponseEntity.status(errorCode.httpStatus)
            .body(Response.withError(errorCode, exception.message))
    }
}